ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/results/tracing/vm/VmTraceGenerator.java

```diff
@@ -0,0 +1,198 @@
+   *
+   * @param index index of the current frame in the trace
+   * @param transactionTrace the transaction trace
+   * @param traceFrame the current trace frame
+   * @param parentTraces the queue of parent traces
+   */
+  private static void addFrame(
+      final AtomicInteger index,
+      final TransactionTrace transactionTrace,
+      final TraceFrame traceFrame,
+      final Deque<VmTrace> parentTraces) {
+    if ("STOP".equals(traceFrame.getOpcode())) {
+      return;
+    }
+
+    VmTrace newSubTrace;
+    VmTrace currentTrace = parentTraces.getLast();
+
+    // set smart contract code
+    currentTrace.setCode(traceFrame.getMaybeCode().orElse(new Code()).getBytes().getHexString());
+    final int nextFrameIndex = index.get() + 1;
+    // retrieve next frame if not last
+    final Optional<TraceFrame> maybeNextFrame =
+        transactionTrace.getTraceFrames().size() > nextFrameIndex
+            ? Optional.of(transactionTrace.getTraceFrames().get(nextFrameIndex))
+            : Optional.empty();
+    final Op op = new Op();
+    // set gas cost and program counter
+    op.setCost(traceFrame.getGasCost().orElse(Gas.ZERO).toLong());
+    op.setPc(traceFrame.getPc());
+
+    final Ex ex = new Ex();
+    // set gas remaining
+    ex.setUsed(
+        traceFrame.getGasRemaining().toLong() - traceFrame.getGasCost().orElse(Gas.ZERO).toLong());
+
+    final boolean isPreviousFrameReturnOpCode =
+        index.get() != 0
+            ? Optional.ofNullable(transactionTrace.getTraceFrames().get(index.get() - 1))
+                .map(frame -> "RETURN".equals(frame.getOpcode()))
+                .orElse(false)
+            : false;
+
+    // set memory if memory has been changed by this operation
+    if (traceFrame.isMemoryWritten() && !isPreviousFrameReturnOpCode) {
+      maybeNextFrame
+          .flatMap(TraceFrame::getMemory)
+          .filter(memory -> memory.length > 0)
+          .map(Trace::dumpAndTrimTrailingZeros)
+          .map(Mem::new)
+          .ifPresent(ex::setMem);
+    }
+
+    // set push from stack elements if some elements have been produced
+    if (traceFrame.getStackItemsProduced() > 0 && maybeNextFrame.isPresent()) {
+      final Bytes32[] stack = maybeNextFrame.get().getStack().orElseThrow();
+      if (stack != null && stack.length > 0) {
+        IntStream.range(0, traceFrame.getStackItemsProduced())
+            .forEach(
+                i -> {
+                  final BytesValue value =
+                      BytesValues.trimLeadingZeros(stack[stack.length - i - 1]);
+                  ex.addPush(value.isEmpty() || value.isZero() ? "0x0" : value.toShortHexString());
+                });
+      }
+    }
+
+    if ("CALL".equals(traceFrame.getOpcode())) {
+      maybeNextFrame.ifPresent(
+          nextFrame -> op.setCost(nextFrame.getGasRemaining().toLong() + op.getCost()));
+      newSubTrace = new VmTrace();
+      parentTraces.addLast(newSubTrace);
+      ex.addPush("0x1");
```

Comment:
I am not sure. But for a reason i ignore there is always this value in `push` field for a trace of a `CALL` operation.

Reply:
do you think it is worth investigating more ?

Reply:
Yes, the current tests are not at all comprehensive, so we can't just assume that because all of the current tests produce this output that this is a generic feature.  We need to figure out where this value is coming from and make the code generic.

Reply:
Looks like you just need to check what's actually on the stack: https://github.com/hyperledger/besu/blob/5ecd562a46f39609b6a3cf8690eccec2882b8a09/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/vm/AbstractCallOperation.java#L221-L225

Reply:
Done.

Reply:
https://github.com/hyperledger/besu/pull/4/commits/90fa8b9a88461387177ec212d394a8153217c92d